source/common/jwt_authn/extractor.cc

```diff
@@ -0,0 +1,138 @@
+#include "common/jwt_authn/extractor.h"
+
+#include "common/common/utility.h"
+#include "common/http/headers.h"
+#include "common/http/utility.h"
+
+using ::Envoy::Http::LowerCaseString;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtAuthentication;
+
+namespace Envoy {
+namespace JwtAuthn {
+namespace {
+
+// A base JwtLocation object to store token and specified_issuers.
+class JwtLocationBase : public JwtLocation {
+public:
+  JwtLocationBase(const std::string& token, const std::unordered_set<std::string>& issuers)
+      : token_(token), specified_issuers_(issuers) {}
+
+  // Get the token string
+  const std::string& token() const override { return token_; }
+
+  // Check if an issuer has specified the location.
+  bool isIssuerSpecified(const std::string& issuer) const override {
+    return specified_issuers_.find(issuer) != specified_issuers_.end();
+  }
+
+private:
+  // Extracted token.
+  std::string token_;
+  // Stored issuers specified the location.
+  const std::unordered_set<std::string>& specified_issuers_;
```

Comment:
Are you sure this won't be dangling reference?

Reply:
Yes, I am sure.   Extractor will be embedded in the FilterConfig object which is part of Listener filter chain,   Token array is part of request data, it last usage should be after JwtAuthn::Verification is done to remove the token.   Extractor life span should exceed the request data lift span.